[RFC] tests/ftests: synchronize between github runners

kamalesh-babulal opened the pull request:
If a VM is shared between multiple github runners, there are chances of
each other stomping over other's run, if executed parallelly.  To avoid the
race between the runners, introduce lock file, that gets acquired (created)
when `ftest.sh` starts and gets removed by `ftest-nocontainer.sh`, this
ensures that both test cases are executed before other runners, that's
waiting for their chance to run.  A runner would wait for 10 minutes before
re-trying to run.  At the max, a runner would wait for 50 minutes (5 retries)
before giving up.

drakenclimber commented on tests/ftests/ftests-nocontainer.sh:
```diff
@@ -21,10 +37,10 @@ fi
 PATH="$PATH:$(pwd)"
 export PATH
-popd || exit $AUTOMAKE_HARD_ERROR
+popd || cleanp $AUTOMAKE_HARD_ERROR
```
typo - should be cleanup :)

kamalesh-babulal replied:
Sorry, it should have been `cleaup`. Will fix it up in the next iteration.

drakenclimber commented on tests/ftests/ftests-nocontainer.sh:
```diff
@@ -6,6 +6,22 @@ AUTOMAKE_HARD_ERROR=99
 
 START_DIR=$PWD
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+RUNNER_LOCK_FILE="/var/run/libcgroup/github-runner.lock"
```
I can see the lock file being needed for `ftests-nocontainer.sh` since the tests are operating on the "bare-metal" `/sys/fs/cgroup`.  But is it needed for `ftests.sh`?  `ftests.sh` generates a random container name for each run, so I don't think container foo should collide with container bar.

https://github.com/libcgroup/libcgroup/blob/main/tests/ftests/ftests.sh#L17

drakenclimber replied:
Or are we doing this because the VMs are underpowered and they can't support simultaneous test runs?

kamalesh-babulal replied:
@drakenclimber Thank you so much for the review.  Yes, it's because of the underpowered VMs.

drakenclimber commented on tests/ftests/ftests.sh:
```diff
@@ -4,6 +4,14 @@
+# each other's run.
+LIBCGROUP_RUN_DIR="/var/run/libcgroup/"
+RUNNER_LOCK_FILE="/var/run/libcgroup/github-runner.lock"
+RUNNER_SLEEP_SECS=600		# sleep for 10 minutes
```
10 minutes feels like a long wait time to me, but I haven't thought about it much :/

kamalesh-babulal replied:
An average run takes between 10 to 15 minutes to complete, hence the lower bound of 10 minutes.

drakenclimber commented:
Looks like a cool idea to me.  I added a few questions, but nothing major.

It does feel weird that `ftests.sh` grabs the lock and `ftests-nocontainer.sh` releases the lock.  Seems like a weird coupling between the two.

kamalesh-babulal commented:
**V2 Changes**
- Reduce the timeout to 5 minutes and increases the max tries to 10 (suggested by @drakenclimber)
- Introduce timeout for races between `apt` operations.

drakenclimber commented:
I have a bad feeling that splitting the lock/unlock into separate files will bite us.  I've reworked the locking patch into this:
https://github.com/drakenclimber/libcgroup/commit/68a4fba70a658b036387bf4fb1664842126e4f93

What do you think, @kamalesh-babulal?

kamalesh-babulal commented:
> I have a bad feeling that splitting the lock/unlock into separate files will bite us. I've reworked the locking patch into this:
https://github.com/drakenclimber/libcgroup/commit/68a4fba70a658b036387bf4fb1664842126e4f93

@drakenclimber I liked your approach, it's cleaner.  `shellcheck`  reported a couple of warnings on the `ftest-wrapper.sh` (ran it locally) and I have fixed them, please feel free to include them:
```
diff --git a/tests/ftests/ftests-wrapper.sh b/tests/ftests/ftests-wrapper.sh
index 10e1c04..5a76c5d 100755
--- a/tests/ftests/ftests-wrapper.sh
+++ b/tests/ftests/ftests-wrapper.sh
@@ -26,10 +26,12 @@ if [ "$START_DIR" != "$SCRIPT_DIR" ]; then
        cp "$SCRIPT_DIR"/*.py "$START_DIR"
 fi
 
-if [ -d ../../src/python/build/lib.* ]; then
-       pushd ../../src/python/build/lib.*
-       export PYTHONPATH="$PYTHONPATH:$(pwd)"
-       popd
+PYTHON_LIBRARY_PATH="../../src/python/build/lib.*"
+if [ -d "$PYTHON_LIBRARY_PATH" ]; then
+       pushd "$PYTHON_LIBRARY_PATH" || cleanup $AUTOMAKE_HARD_ERROR
+       PYTHONPATH="$PYTHONPATH:$(pwd)"
+       export PYTHONPATH
+       popd || cleanup $AUTOMAKE_HARD_ERROR
 fi
 
 # If other runners are running then the file exists
@@ -66,7 +68,7 @@ PATH="$PATH:$(pwd)"
 export PATH
 popd || cleanup $AUTOMAKE_HARD_ERROR
 
-sudo PATH=$PATH PYTHONPATH=$PYTHONPATH ./ftests.py -l 10 -s "sudo" \
+sudo PATH="$PATH" PYTHONPATH="$PYTHONPATH" ./ftests.py -l 10 -s "sudo" \
        -L "$START_DIR/ftests-nocontainer.py.sudo.log" --no-container -n Libcg"$RANDOM"
 RET3=$?
 
```
in line 89:
```
   89 if [[ $RET1 -eq 0 ]] || [[ $RET2 -eq 0 ]] || [[ $RET3 -eq 0 ]]; then
   90         # at least one test passed and _no_ tests failed
   91         cleanup 0
   92 fi
```

consider this if any of the `RET?` is set to some non-zero value (failed test count), the test will
return a false positive.  A check to ensure that `RET` is `$AUTOMAKE_SKIPPED` when it's>0 is
required.  I am a little  biased by the way you handled it in `ftest-nocontainer.sh`

drakenclimber commented:
> @drakenclimber I liked your approach, it's cleaner. `shellcheck` reported a couple of warnings on the `ftest-wrapper.sh` (ran it locally) and I have fixed them, please feel free to include them:
> 

Thanks.  Will do.

> 
> in line 89:
> 
> ```
>    89 if [[ $RET1 -eq 0 ]] || [[ $RET2 -eq 0 ]] || [[ $RET3 -eq 0 ]]; then
>    90         # at least one test passed and _no_ tests failed
>    91         cleanup 0
>    92 fi
> ```
> 
> consider this if any of the `RET?` is set to some non-zero value (failed test count), the test will return a false positive. A check to ensure that `RET` is `$AUTOMAKE_SKIPPED` when it's>0 is required. I am a little biased by the way you handled it in `ftest-nocontainer.sh`

Oh, good catch.  I forgot that it returns the `failed_cnt`.  The logic in `ftest-nocontainer` worked well for two tests, but it was unwieldy for three.  Thus the change.

Let me think about it a bit.  I'll post an update today.

drakenclimber commented:
Alright, here's the update:
https://github.com/drakenclimber/libcgroup/commit/cffcbaad7cdca7353d00804d152bf82485d6548f

It was much easier to add the ftests-nocontainer `RET` logic than I thought.  Good call.

I did have to modify the PYTHON_LIBRARY_PATH logic a little bit.  Since we're wildcarding the path, we can't put it in quotes.

kamalesh-babulal commented:
> Alright, here's the update:
https://github.com/drakenclimber/libcgroup/commit/cffcbaad7cdca7353d00804d152bf82485d6548f
It was much easier to add the ftests-nocontainer RET logic than I thought. Good call.

@drakenclimber Looks good to me :)

> I did have to modify the PYTHON_LIBRARY_PATH logic a little bit. Since we're wildcarding the path, we can't put it in quotes.

`shellcheck` is unhappy with the way we are using glob:
```
# shellcheck ftests-wrapper.sh 

In ftests-wrapper.sh line 29:
PYTHON_LIBRARY_PATH=../../src/python/build/lib.*
                    ^--------------------------^ SC2125 (warning): Brace expansions and globs are literal in assignments. Quote it or use an array.


In ftests-wrapper.sh line 30:
if [ -d $PYTHON_LIBRARY_PATH ]; then
        ^------------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
if [ -d "$PYTHON_LIBRARY_PATH" ]; then


In ftests-wrapper.sh line 31:
        pushd $PYTHON_LIBRARY_PATH || cleanup $AUTOMAKE_HARD_ERROR
              ^------------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
        pushd "$PYTHON_LIBRARY_PATH" || cleanup $AUTOMAKE_HARD_ERROR

For more information:
  https://www.shellcheck.net/wiki/SC2125 -- Brace expansions and globs are li...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

kamalesh-babulal commented:
`shellcheck` seems to be okay with:
````
diff --git a/tests/ftests/ftests-wrapper.sh b/tests/ftests/ftests-wrapper.sh
index b7fd64b..95ec0cd 100755
--- a/tests/ftests/ftests-wrapper.sh
+++ b/tests/ftests/ftests-wrapper.sh
@@ -26,9 +26,9 @@ if [ "$START_DIR" != "$SCRIPT_DIR" ]; then
        cp "$SCRIPT_DIR"/*.py "$START_DIR"
 fi
 
-PYTHON_LIBRARY_PATH=../../src/python/build/lib.*
-if [ -d $PYTHON_LIBRARY_PATH ]; then
-       pushd $PYTHON_LIBRARY_PATH || cleanup $AUTOMAKE_HARD_ERROR
+PYTHON_LIBRARY_PATH=(../../src/python/build/lib*)
+if [ -d  "${PYTHON_LIBRARY_PATH[0]}" ]; then
+       pushd "${PYTHON_LIBRARY_PATH[0]}" || cleanup $AUTOMAKE_HARD_ERROR
        PYTHONPATH="$PYTHONPATH:$(pwd)"
        export PYTHONPATH
        popd || cleanup $AUTOMAKE_HARD_ERROR

drakenclimber commented:
Resolved in pull request #310.
